### PR TITLE
Feature/PO-14: Block Creation

### PR DESCRIPTION
**Pull request by javierdiaz72**
The intention of this PR is to add the functionality required for stakeholders to generate new blocks.

### CODE REVIEW

File: src/main/scala/io/iohk/praos/domain/ConsensusResolver.scala [hunk 1]
```diff
@@ -21,7 +21,7 @@ object ConsensusResolver {
   private def join(baseBlockchain: Blockchain, possibleExtension: Blockchain): Option[Blockchain] = {
     possibleExtension match {
       case Nil => Some(baseBlockchain)
-      case possibleExtensionHead :: _ if baseBlockchain.lastOption.map(_.blockHash) == possibleExtensionHead.state =>
+      case possibleExtensionHead :: _ if baseBlockchain.lastOption.map(block => blockHash(block)) == possibleExtensionHead.value.state =>
```
Comment (javierdiaz72) [inline src/main/scala/io/iohk/praos/domain/ConsensusResolver.scala]: The lambda expression could be simplified to `blockHash(_)`.

File: src/main/scala/io/iohk/praos/App.scala [hunk 2]
```diff
@@ -1,8 +1,9 @@
 package io.iohk.praos
 
 import io.iohk.praos.domain._
-import io.iohk.praos.crypto.generateNewRandomValue
+import io.iohk.praos.crypto.{SignerImpl, VerifiableRandomFunctionStubImpl, generateNewRandomValue}
 import io.iohk.praos.util.Logger
+//import io.iohk.praos.util.TransactionsGenerator
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/App.scala]: Minor detail: Should we remove this comment if it's not used?

File: src/main/scala/io/iohk/praos/crypto/package.scala [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
   type RandomValue = Int
 
   // Dummy implementation to simulate a public-private key schema.
   val keyLength = 32
 
-  private def genKey = (s: RandomValue, length: Int) => ByteString(Array.fill(length)(s.toByte))
+  def genKey = (s: RandomValue, length: Int) => ByteString(Array.fill(length)(s.toByte))
 
-  private def XOR(x: ByteString, y: ByteString): ByteString =
+  def XOR(x: ByteString, y: ByteString): ByteString =
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/crypto/package.scala]: As this function is only used in the CipherStubImpl, maybe it should be better to move it there and have it private? If that's not the case maybe we could create a `utils.ByteUtils` object with this auxiliar functions.

File: src/main/scala/io/iohk/praos/crypto/SignerImpl.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+object SignerImpl extends Signer {
+  override def signedWith[T <: Serializable](data: T, privateKey: Key): Signed[T] = {
+    // FIXME: Hasher, cipher and signature provider are hard-coded for simplicity.
+    val publicKey = getPublicKeyFromPrivateKey(privateKey)
+    val hasher = PredefinedHasher("MD5")
+    val serializedData = ByteString(data.toString)
+    val hashedData = hasher.hash(serializedData)
+    val encryptedData = CipherStubImpl.encryptWith(hashedData, publicKey)
+    val signature = SignatureProviderImpl.getSignature(encryptedData, privateKey)
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/crypto/SignerImpl.scala]: Could we instead `SignatureProviderImpl.getSignature(serializedData, privateKey)`? Maybe I'm missing something but if this hashing and encryption is needed maybe it should be better for it to be done on the `SignatureProvider`.

File: src/main/scala/io/iohk/praos/crypto/package.scala [hunk 5]
```diff
@@ -73,34 +42,8 @@ package object crypto {
...
-    * @param algorithm  A pre-defined algorithm (e.g. "MD5").
-    * @note  Precondition: The given algorithm is supported.
-    */
-  case class PredefinedHasher(algorithm: String) extends Hasher {
-    override def hash(message: ByteString): Digest = {
-      ByteString(java.security.MessageDigest.getInstance(algorithm).digest(message.toArray))
-    }
-  }
+  // TODO: Change to a realistic value.
+  def seedNonce: Seed = 123
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/crypto/package.scala]: Minor detail: If I understood correctly this value won't change, right? If that's the case we should probably change it to `val SeedNonce: Seed = 123`

File: src/main/scala/io/iohk/praos/domain/BlockFactory.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package io.iohk.praos.domain
+
+import io.iohk.praos.crypto._
+
+
+case class BlockFactory(signer: Signer, vrf: VerifiableRandomFunction) {
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: ~Maybe it would be better to move this to the `utils` package?~
Thinking better about it, maybe utils isn't the best place, this isn't exactly as with `TransactionsGenerator` as it won't eventually disappear as with it, so maybe we can leave it here for now.

Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: Maybe we should rename this to `BlockGenerator` or rename `TransactionsGenerator` to `TransactionsFactory` for consistency? What do you think @mkatixlabs ?

File: src/main/scala/io/iohk/praos/domain/StakeholderState.scala [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import io.iohk.praos.crypto.{Hasher, Key}
+
+
+object StakeholderState {
+  // This function corresponds to the initStakeholderStake function in the Praos Formalization.
+  def initialize(privateKey: Key, genesisBlock: GenesisBlock): StakeholderState =
+    StakeholderState(privateKey, genesisBlock, genesisBlock, Blockchain(), None, List.empty)
+}
+
+case class StakeholderState(
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/StakeholderState.scala]: Why is this needed? Maybe I'm missing something but it doesn't seem to be currently used. Also check `BlockchainState` which already has part of the information from this class.

File: src/main/scala/io/iohk/praos/domain/BlockFactory.scala (hunk 6 above)
Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: Maybe this should start using the `TransactionsGenerator`? It shouldn't be too complicated but if it is we could create a separate task to make the change.

File: src/main/scala/io/iohk/praos/crypto/package.scala [hunk 8]
```diff
@@ -1,21 +1,22 @@
...
+
   type Key = ByteString
   type KeyPair = (Key, Key)
   type RandomValue = Int
 
   // Dummy implementation to simulate a public-private key schema.
   val keyLength = 32
 
-  private def genKey = (s: RandomValue, length: Int) => ByteString(Array.fill(length)(s.toByte))
+  def genKey = (s: RandomValue, length: Int) => ByteString(Array.fill(length)(s.toByte))
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/crypto/package.scala]: Minor detail: Why is this function no longer `private`? It seems to only be used in this file.

File: src/main/scala/io/iohk/praos/domain/ConsensusResolver.scala (hunk 1 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/ConsensusResolver.scala]: Resolved.

File: src/main/scala/io/iohk/praos/App.scala (hunk 2 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/App.scala]: Resolved.

File: src/main/scala/io/iohk/praos/crypto/SignerImpl.scala (hunk 4 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/SignerImpl.scala]: AFAIK, the hashing of the data is normally required for signing something, whereas the encryption part is normally optional.

On another hand, `SignatureProvider` should disappear as the implementation evolves, since it only implements a very simple way to obtain the public key from the signature (basically by embedding the PK into the signature) and is of no much value in itself as an abstraction.

File: src/main/scala/io/iohk/praos/crypto/package.scala (hunk 3 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/package.scala]: Correct, `XOR` is currently used in `CipherStubImpl` only. However, I think it's a good idea to make it available to other potential users by creating the `utils.ByteUtils` object.

File: src/main/scala/io/iohk/praos/crypto/package.scala (hunk 5 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/package.scala]: Resolved.

File: src/main/scala/io/iohk/praos/domain/BlockFactory.scala (hunk 6 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: In my experience, "Generator" is used more for representing things like RNGs or fake data producers like `TransactionGenerator`. At least to me, generators and factories are different concepts (in a subtle way though).

File: src/main/scala/io/iohk/praos/domain/StakeholderState.scala (hunk 7 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/StakeholderState.scala]: Correct, removed.

File: src/main/scala/io/iohk/praos/crypto/package.scala (hunk 8 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/package.scala]: Correct, fixed.

File: src/main/scala/io/iohk/praos/domain/BlockFactory.scala (hunk 6 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: IMHO, it's not a good idea to couple the transactions generation to the way blocks are created by the factory. A block should receive a list of the transactions to be included in the block, how these transactions are created should be a separate matter.

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockFactory.scala:6]: As discussed with Nico and Mirko, the current design is fine.

File: src/main/scala/io/iohk/praos/crypto/SignerImpl.scala (hunk 4 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/SignerImpl.scala]: As discussed with Nico, `SignatureProvider.splitSignature` will be renamed to `SignatureProvider.getPublicKeyFromSignature` and return only the public key. Agreed that this class will eventually disappear.

File: src/main/scala/io/iohk/praos/domain/BlockSigner.scala [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package io.iohk.praos.domain
+
+import akka.util.ByteString
+import io.iohk.praos.crypto._
+
+
+object BlockSigner {
+  def signedWith(block: Block, privateKey: Key): Block = {
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/BlockSigner.scala]: I'm not sure whether having to pass a block with a fake signature to this function. Maybe we could have both `UnsignedBlock` and `Block` classes, what do you think?

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockSigner.scala]: Honestly I don't like having an "empty" or "fake" signature for "not yet signed" blocks either. However, if you convience Alan V to use inheritance to subclass Block from a new UnsignedBlock class, I'll agree with you :)

Comment (ntallar) [reply src/main/scala/io/iohk/praos/domain/BlockSigner.scala]: I don't like that either :stuck_out_tongue:
I was thinking that maybe we could have a `Block` contain an `UnsignedBlock`, WDYT?

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockSigner.scala]: Yes, but we would basically be back to the point of having the Signed case class :)

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/BlockSigner.scala]: After discussing this with Nico and Alan V, we will revert these changes and stick to the `UnsignedBlock`/`Block`/`Signed[T]` modeling as defined by the Praos Formalization.

File: src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala [hunk 10]
```diff
@@ -2,33 +2,29 @@ package io.iohk.praos.crypto
...
-
-object VrfProof {
-  // TODO: When the VrfProof be well defined, change to the true random length
-  val randomLength = 32
-}
 
 trait VerifiableRandomFunction {
 
-  def prove(privateKey: Key, randomSeed: RandomValue): VrfProof
+  type VrfProof = ByteString
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala]: Minor detail: Maybe we should move this to `object VerifiableRandomFunction`?

File: src/main/scala/io/iohk/praos/domain/package.scala [hunk 11]
```diff
@@ -37,12 +39,17 @@ package object domain {
 
   type SlotNumber = Int
 
+  type Block = Signed[UnsignedBlock]
+  def Block(unsignedBlock: UnsignedBlock, signature: Signature) = Signed[UnsignedBlock](unsignedBlock, signature)
```
Comment (ntallar) [inline src/main/scala/io/iohk/praos/domain/package.scala]: Maybe we could move this and the `blockHash` function to a separate `object Block` in a separate file?

**Review by ntallar — COMMENTED** [review]
Apart from some minor comments (and the merge conflicts), LGTM!

File: src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala (hunk 10 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala]: IMHO, the implementation of the `VrfProof` type should not depend on an specific implementation of the `VerifiableRandomFunction` trait.

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala]: I misunderstood Nico's argument. Moving the type definition to a companion object.

Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/crypto/VerifiableRandomFunction.scala]: Resolved

File: src/main/scala/io/iohk/praos/domain/package.scala (hunk 11 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/package.scala]: Resolved.

**Review by ntallar — APPROVED** [review]
LGTM!
